tests/message-box-test.lisp

```diff
@@ -0,0 +1,121 @@
+(defpackage :sento.message-box-test
+  (:use :cl :fiveam :cl-mock :sento.actor :sento.future)
+  (:shadow #:! #:?)
+  (:import-from #:miscutils
+                #:assert-cond
+                #:await-cond
+                #:filter)
+  (:import-from #:timeutils
+                #:ask-timeout)
+  (:import-from #:sento.messageb
+                #:message-box/bt
+                #:submit
+                #:no-result
+                #:queue-thread
+                #:stop)
+  (:import-from #:ac
+                #:actor-of))
+
+(in-package :sento.message-box-test)
+
+(def-suite message-box-tests
+  :description "message-box tests"
+  :in sento.tests:test-suite)
+
+(in-suite message-box-tests)
+
+
+(defun wait-while-thread-will-die (msgbox &key (timeout 10))
+  (let ((wait-until (+ (get-internal-real-time) (* timeout
+                                                   internal-time-units-per-second))))
+    (with-slots (queue-thread)
+        msgbox
+      (loop while (bt2:thread-alive-p queue-thread)
+            do (sleep 0.1)
+               (when (< wait-until
+                        (get-internal-real-time))
+                 (error "Thread didn't die in ~A seconds."
+                        timeout))))))
+
+
+(test bt-box-resurrects-thread-after-error
+  "Tests that if an error happends during message processing, a thread will remain running."
+
+  (let ((box (make-instance 'message-box/bt
+                            :name "foo")))
+    (unwind-protect
+         (progn
+           (let ((first-reply
+                   (submit box "The Message"
+                           t
+                           ;; Don't wait for result here, because we are
+                           ;; intentionally raise error here and will never
+                           ;; return a result:
+                           nil
+                           (list (lambda (msg)
+                                   (declare (ignore msg))
+                                   (handler-bind ((serious-condition #'abort))
+                                     (error "Die, thread, die!")))))))
+             (is (equal first-reply
+                        'no-result)))
+
+           (wait-while-thread-will-die box)
+
+           (let ((result (handler-case
+                             (submit box "The Message" t 1
+                                     (list (lambda (msg)
+                                             (reverse msg))))
+                           (ask-timeout ()
+                             :timeout))))
+             (is (string= "egasseM ehT" result))))
+      
+      ;; Cleanup a thread:
+      (stop box t))))
+
+
+(test bt-box-resurrects-thread-after-abort-if-handler-catches-all-signals
```

Comment:
Probably we should leave only this test and to remove the first one?

Reply:
I need to closely look through the tests. Didn't yet have enough time. But generally the tests should cover the behavior of the thread handling and be a spec for how it should work.
So we should check cases where things are running normally, thread doesn't die. And also where the thread dies for whatever reason. The reasons may be many, but the gist probably is something that breaks the pop-loop.

Reply:
I see two ways of loosing the thread:

1. Because of non-local exit from the processing loop (as the result of some error signal or restart invocation).
2. Because somebody intentionally terminated the thread.

I think it will be possible to detect the first case and instead of checking if thread is alive each time actor receives a message we will mark it as "thread-died" on unwind protect. Checking slot "thread-died" should be faster than checking if thread is alive.

And probably we should ignore the second case, because if somebody decided to kill the thread it his responsiblity to recreate actors.

Reply:
That's OK.